internal/factsengine/factsengine.go

```diff
@@ -0,0 +1,201 @@
+	"golang.org/x/sync/errgroup"
+)
+
+type FactsEngine struct {
+	agentID             string
+	factsEngineService  string
+	factGatherers       map[string]gatherers.FactGatherer
+	factsServiceAdapter adapters.Adapter
+}
+
+func NewFactsEngine(agentID, factsEngineService string) *FactsEngine {
+	return &FactsEngine{
+		agentID:            agentID,
+		factsEngineService: factsEngineService,
+		factGatherers: map[string]gatherers.FactGatherer{
+			gatherers.CorosyncFactKey: gatherers.NewCorosyncConfGatherer(),
+		},
+	}
+}
+
+func (c *FactsEngine) Subscribe() error {
+	log.Infof("Subscribing agent %s to the facts gathering reception service on %s", c.agentID, c.factsEngineService)
+	//RabbitMQ adapter exists only by now
+	factsServiceAdapter, err := adapters.NewRabbitMQAdapter(c.factsEngineService)
+	if err != nil {
+		return err
+	}
+
+	c.factsServiceAdapter = factsServiceAdapter
+	log.Infof("Subscription to the facts engine by agent %s in %s done", c.agentID, c.factsEngineService)
+
+	return nil
+}
+
+func (c *FactsEngine) Unsubscribe() error {
+	log.Infof("Unsubscribing agent %s from the facts engine service", c.agentID)
+	if err := c.factsServiceAdapter.Unsubscribe(); err != nil {
+		return err
+	}
+
+	log.Infof("Unsubscribed properly")
+
+	return nil
+}
+
+func (c *FactsEngine) Listen(ctx context.Context) error {
+	var err error
+
+	log.Infof("Listening for facts gathering events...")
+	defer func() {
+		err = c.Unsubscribe()
+		if err != nil {
+			log.Errorf("Error during unsubscription: %s", err)
+		}
+	}()
+
+	if err := c.factsServiceAdapter.Listen(c.agentID, c.handleRequest); err != nil {
+		return err
+	}
+
+	<-ctx.Done()
+
+	return err
+}
+
+func (c *FactsEngine) handleRequest(request []byte) error {
+	factsRequests, err := parseFactsRequest(request)
+	if err != nil {
+		log.Errorf("Invalid facts request: %s", err)
+		return err
+	}
+
+	gatheredFacts, err := gatherFacts(factsRequests, c.factGatherers)
+	if err != nil {
+		log.Errorf("Error gathering facts: %s", err)
+		return err
+	}
+
+	if err := c.publishFacts(gatheredFacts); err != nil {
+		log.Errorf("Error publishing facts: %s", err)
+		return err
+	}
+
+	return nil
+}
+
+func gatherFacts(groupedFactsRequest *gatherers.GroupedFactsRequest, factGatherers map[string]gatherers.FactGatherer) (gatherers.FactsResult, error) {
+	factsResults := gatherers.FactsResult{
+		ExecutionID: groupedFactsRequest.ExecutionID,
+	}
+	log.Infof("Starting facts gathering process")
+
+	// Gather facts asynchronously
+	g := new(errgroup.Group)
+
+	for gathererType, f := range groupedFactsRequest.Facts {
+		factsRequest := f
+
+		gatherer, exists := factGatherers[gathererType]
+		if !exists {
+			log.Errorf("Fact gatherer %s does not exist", gathererType)
+			continue
+		}
+
+		// Execute the fact gathering asynchronously and in parallel
+		g.Go(func() error {
+			newFacts, err := gatherer.Gather(factsRequest)
+			if err == nil {
+				factsResults.Facts = append(factsResults.Facts, newFacts[:]...)
```

Comment:
appending an array in a goroutine like that can lead to some race conditions or unexpected behavior, we should use a channel that the goroutine uses to send the new facts, and have a channel loop in the gather facts function that collect this result and append in a single place, in a single routine to the facts inside the factsResult struct.

At first glance this can work, but there is the risk, there are multiple goroutine involved, and they can lead to a race

Reply:
Good point!
Changed to use a buffered channel

Reply:
I don't know If I am clear about that, but we need another goroutine and the waitgroup, the other goroutine will defer the channel closing when all the routines sent the payload on channel.